File: 41/solution.rb
# one-pass

# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
    n = nums.size
    n.times do |i|
        while nums[i] > 0 && nums[i] <= n && nums[i] != nums[nums[i] - 1]
            rightIdx = nums[i] - 1
            nums[rightIdx], nums[i] = nums[i], nums[rightIdx]
        end
    end
    
    nums.each_with_index do |num, i|
        return i + 1 if num != i + 1
    end
    
    n + 1
end

File: 41/solution.py
# one-pass

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        for i in range(n):
            while nums[i] > 0 and nums[i] <= n and nums[i] != nums[nums[i] - 1]:
                rightIdx = nums[i] - 1
                nums[rightIdx], nums[i] = nums[i], nums[rightIdx]
                
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1
            
        return n + 1
        
